Add metrics logging and an Activity page to show requests

g2mt opened the pull request:
This PR adds an Activity page inspired by OpenRouter to show requests sent to the server. It also adds the config settings `metricsMaxInMemory`, `metricsLogPath`, `metricsUseServerResponse`.

<img width="1201" height="401" alt="llama-swap" src="https://github.com/user-attachments/assets/e1c85b3c-965e-46be-b88c-b3af88da6d52" />

The `metricsUseServerResponse` config controls logging behavior. By default, metricsUseServerResponse is false, meaning the token processing metrics are retrieved from the llama-server eval logs, similar to the existing Log Stats section. If set to true then the final response JSON from the llama-server will be parsed to obtain token usage statistics. Retrieving usage from streaming responses is not implemented.



I used Roo Code, aider with mostly Kimi-K2 to assist in this PR.



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced real-time tracking and display of token usage metrics, including input/output tokens, processing speed, and duration.
  * Added a new "Activity" page in the user interface to view recent metrics.
  * Provided an API endpoint and event stream for accessing metrics data.

* **Improvements**
  * Added new configuration option to control the maximum number of stored metrics in memory.
  * Enhanced API and UI to display and manage token usage statistics.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

mostlygeek commented:
I took a look over this PR and there are some design issues prevent accepting the contribution: 

- no support streaming requests. 
- usage of `httptest.NewRecorder` to buffer the response to extract the token information 
- introduction of state (log file) into llama-swap. I'm not entirely against this but llama-swap has been a stateless proxy to date and I would like to be thoughtful before changing this key aspect of it. 

PR #184 introduced an event bus into llama-swap. This powers the real time updates in the UI. I think this is the way forward for future metrics collecting.

g2mt commented:
@mostlygeek 

> I took a look over this PR and there are some design issues that prevent accepting the contribution:
> 
> * no support for streaming requests.
> * usage of `httptest.NewRecorder` to buffer the response to extract the token information

I will add separate classes for parsing the SSE stream, and parsing HTTP response.

> * introduction of state (log file) into llama-swap. I'm not entirely against this but llama-swap is a stateless proxy and I would like to be thoughtful about changing this key aspect of it.

Got it. For now, I removed the log file functionality.

> PR #184 introduced an event bus into llama-swap. This powers the real time updates in the UI. I think this is the way forward for future metrics collecting.

I added an event dispatcher to the metrics parsing struct. I'll go ahead and also add auto-updates to the Activity UI if that's what you meant.

g2mt commented:
@mostlygeek 

I have refactored the HTTP request interception into a separate metrics middleware. It should also handle streaming responses now.

mostlygeek commented:
A few more changes: 

- The changes to process.go, processgroup.go and the appropriate tests should be reverted. They do not need to be coupled to the new metrics functionality. 
- Take a look at `APIProvider.apiiEventSource`, llama-swap uses this single SSE channel to keep the state of the backend/UI in sync.  I removed `proxyEventSource` and `upstreamEventSource` from the code today as they're not used for anything anymore.

g2mt commented:
> @g2mt question: Since the metrics are pulled directly out of the response body is log parsing still required?

Log parsing was implemented because the Log Stats section in the Models tab parses the upstream logs to get usage information. For that I implemented both response interception and log interception for parity.

Would you rather just keep the response interception and revert those changes?

g2mt commented:
> * Take a look at `APIProvider.apiiEventSource`, llama-swap uses this single SSE channel to keep the state of the backend/UI in sync. 

I'm not sure what you meant by this. The activity stream is bound to the `/api/metrics/stream`, which is the same as how logging is done.

g2mt commented:
@mostlygeek I have removed the raw log parsing functionality and moved the metrics streaming handler to the event stream.

g2mt commented on proxy/proxymanager.go:
```diff
@@ -149,18 +153,21 @@ func (pm *ProxyManager) setupGinEngine() {
-	pm.ginEngine.POST("/v1/audio/speech", pm.proxyOAIHandler)
-	pm.ginEngine.POST("/v1/audio/transcriptions", pm.proxyOAIPostFormHandler)
+	pm.ginEngine.POST("/v1/audio/speech", mmSetup, pm.proxyOAIHandler, mmFlush)
+	pm.ginEngine.POST("/v1/audio/transcriptions", mmSetup, pm.proxyOAIPostFormHandler, mmFlush)
```
Token information for `/audio/transcriptions` seem to be returned as a normal JSON body similar to the other endpoints. It should be handled correctly as is.

mostlygeek replied:
This code will always exit early for /v1/audio/speech, /v1/audio/transcriptions. 

```go
func (rec *MetricsRecorder) parseAndRecordMetrics(jsonData gjson.Result) {
	if !jsonData.Get("usage").Exists() {
		return
	}
....
```

In v1/audio/transcriptions the `usage` key is never set. In v1/audio/speech the output is a binary audio file. There's no need to incur the extra memory and CPU overhead for these endpoints.

g2mt commented on proxy/metrics_middleware.go:
```diff
@@ -0,0 +1,142 @@
+
+// MetricsMiddlewareFlush uses the writer's recorded HTTP response for metrics processing.
+// The middleware expects the fields modelName,... to be set after setup.
+func MetricsMiddlewareFlush(c *gin.Context) {
```
Fixed.

g2mt commented on proxy/proxymanager.go:
```diff
@@ -404,11 +410,19 @@ func (pm *ProxyManager) proxyOAIHandler(c *gin.Context) {
 		return
 	}
+
+	if writer, ok := c.Writer.(*MetricsResponseWriter); ok {
```
If we're deduplicating the requestedModel extraction then the endpoint with proxyOAIHandler always requires MetricsMiddleware right? Maybe a comment should be added to proxyOAIHandler's function signature to reflect this.

mostlygeek commented on proxy/proxymanager_test.go:
```diff
@@ -165,9 +165,11 @@ func TestProxyManager_SwapMultiProcessParallelRequests(t *testing.T) {
 
 			mu.Lock()
-			var response map[string]string
+			var response map[string]interface{}
```
Oh I see. Nope it's fine to leave it as is.

mostlygeek commented:
Released in v140
